pkgs/applications/misc/k2pdfopt/default.nix

```diff
@@ -11,44 +11,119 @@
+# of creating and applying these patches in this way is that minor updates
+# (esp. security fixes) might still allow these patches to apply successfully.
+#
+# The current approach creates these patches using nix. This has the benefit of
+# easier review and should hopefully be simpler to update in the future.
+
+let
+  # Create a patch against src based on changes applied in patchCommands
+  mkPatch = { name, src, patchCommands }: runCommand "${name}-k2pdfopt.patch" { inherit src; } ''
+    source $stdenv/setup
+    unpackPhase
 
-  src = (fetchzip {
-    url = "http://www.willus.com/k2pdfopt/src/k2pdfopt_v2.51_src.zip";
-    sha256 = "133l7xkvi67s6sfk8cfh7rmavbsf7ib5fyksk1ci6b6sch3z2sw9";
-  });
+    orig=$sourceRoot
+    new=$sourceRoot-modded
+    cp -r $orig/. $new/
 
-  # Note: the v2.51a zip contains only files to be replaced in the v2.50 zip.
-  v251a_src = (fetchzip {
-    url = "http://www.willus.com/k2pdfopt/src/k2pdfopt_v2.51a_src.zip";
-    sha256 = "0vvwblii7kgdwfxw8dzk6jbmz4dv94d7rkv18i60y8wkayj6yhl6";
-  });
+    pushd $new >/dev/null
+    ${patchCommands}
+    popd >/dev/null
 
-  postUnpack = ''
-    cp -r ${v251a_src}/* $sourceRoot
+    diff -Naur $orig $new > $out || true
   '';
 
-  patches = [ ./k2pdfopt.patch ./k2pdfopt-mupdf-1.16.1.patch ];
+  pname = "k2pdfopt";
+  version = "2.53";
+  k2pdfopt_src = fetchzip {
+    url = "http://www.willus.com/${pname}/src/${pname}_v${version}_src.zip";
+    sha256 = "1fna8bg3pascjfc3hmc6xn0xi2yh7f1qp0d344mw9hqanbnykyy8";
+  };
+in stdenv.mkDerivation rec {
+  inherit pname version;
+  src = k2pdfopt_src;
+
+  patches = [
+    ./0001-Fix-CMakeLists.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace willuslib/bmpdjvu.c \
+      --replace "<djvu.h>" "<libdjvu/ddjvuapi.h>"
+  '';
 
   nativeBuildInputs = [ cmake pkgconfig ];
 
   buildInputs =
   let
-    #  The patches below were constructed by taking the files from k2pdfopt in
-    #  the {mupdf,leptonica,tesseract}_mod/ directories, replacing the
-    #  corresponding files in the respective source trees, resolving any errors
-    #  with more recent versions of these depencencies, and running diff.
-    mupdf_modded = mupdf.overrideAttrs (attrs: {
-      patches = attrs.patches ++ [ ./mupdf.patch ]; # Last verified with mupdf 1.16.1
+    # We use specific versions of these sources to match the versions used in the k2pdfopt source
+    mupdf_patch = mkPatch {
+      name = "mupdf";
+      src = fetchurl {
+        url = "https://mupdf.com/downloads/archive/mupdf-1.17.0-source.tar.gz";
+        sha256 = "13nl9nrcx2awz9l83mlv2psi1lmn3hdnfwxvwgwiwbxlkjl3zqq0";
+      };
+      patchCommands = ''
+        cp ${k2pdfopt_src}/mupdf_mod/{filter-basic,font,stext-device,string}.c ./source/fitz/
+        cp ${k2pdfopt_src}/mupdf_mod/pdf-* ./source/pdf/
+      '';
+    };
+    mupdf_modded = mupdf.overrideAttrs ({ patches ? [], ... }: {
+      patches = patches ++ [ mupdf_patch ];
+      # THis function is missing in font.c, see font-win32.c
```

Comment:
THis -> This

Is this only an issue with the patch in k2pdfopt, not in uptream?

Reply:
Thanks. Yes, this appears to be a change k2pdfopt made.